File: resources/demo/sample-solutions/two-sum.cpp
/**
 * Two Sum - C++ Solutions
 * ========================
 * Problem: Given an array of integers and a target, return indices of two numbers
 * that add up to the target.
 */
#include <vector>
#include <unordered_map>
#include <cassert>

// ============================================================
// Solution 1: Hash Table (Optimal)
// Time: O(n), Space: O(n)
// ============================================================
std::vector<int> twoSum(std::vector<int>& nums, int target) {
    // Map from value to its index
    std::unordered_map<int, int> seen;

    for (int i = 0; i < nums.size(); i++) {
        int complement = target - nums[i];

        // Check if complement exists in the map
        auto it = seen.find(complement);
        if (it != seen.end()) {
            return {it->second, i};
        }

        // Store current number and its index
        seen[nums[i]] = i;
    }

    return {};  // No solution found
}

// ============================================================
// Solution 2: Brute Force (for comparison)
// Time: O(n^2), Space: O(1)
// ============================================================
std::vector<int> twoSumBrute(std::vector<int>& nums, int target) {
    for (int i = 0; i < nums.size(); i++) {
        for (int j = i + 1; j < nums.size(); j++) {
            if (nums[i] + nums[j] == target) {
                return {i, j};
            }
        }
    }
    return {};
}

// ============================================================
// Test
// ============================================================
int main() {
    std::vector<int> nums1 = {2, 7, 11, 15};
    auto result1 = twoSum(nums1, 9);
    assert(result1[0] == 0 && result1[1] == 1);

    std::vector<int> nums2 = {3, 2, 4};
    auto result2 = twoSum(nums2, 6);
    assert(result2[0] == 1 && result2[1] == 2);

    std::vector<int> nums3 = {3, 3};
    auto result3 = twoSum(nums3, 6);
    assert(result3[0] == 0 && result3[1] == 1);

    return 0;
}

File: resources/demo/sample-solutions/two-sum.py
"""
Two Sum - Python Solutions
===========================
Problem: Given an array of integers and a target, return indices of two numbers
that add up to the target.
"""

# ============================================================
# Solution 1: Hash Table (Optimal)
# Time: O(n), Space: O(n)
# ============================================================
def two_sum_hash(nums, target):
    """
    Use a hash table to store seen numbers and their indices.
    For each number, check if its complement (target - num) exists.
    """
    seen = {}  # value -> index
    for i, num in enumerate(nums):
        complement = target - num
        if complement in seen:
            return [seen[complement], i]
        seen[num] = i
    return []  # No valid pair found


# ============================================================
# Solution 2: Brute Force (for comparison)
# Time: O(n^2), Space: O(1)
# ============================================================
def two_sum_brute(nums, target):
    """
    Check every pair of numbers. Simple but slow.
    """
    for i in range(len(nums)):
        for j in range(i + 1, len(nums)):
            if nums[i] + nums[j] == target:
                return [i, j]
    return []


# ============================================================
# Solution 3: Two Pass Hash Table
# Time: O(n), Space: O(n)
# ============================================================
def two_sum_two_pass(nums, target):
    """
    First pass: build the hash table.
    Second pass: look for complement.
    Slightly less efficient than one-pass but clearer logic.
    """
    # First pass: build index map
    num_map = {}
    for i, num in enumerate(nums):
        num_map[num] = i

    # Second pass: find complement
    for i, num in enumerate(nums):
        complement = target - num
        if complement in num_map and num_map[complement] != i:
            return [i, num_map[complement]]
    return []


# ============================================================
# Test
# ============================================================
if __name__ == "__main__":
    assert two_sum_hash([2, 7, 11, 15], 9) == [0, 1]
    assert two_sum_hash([3, 2, 4], 6) == [1, 2]
    assert two_sum_hash([3, 3], 6) == [0, 1]
    print("All tests passed!")
